molpipeline/mol2mol/filter.py

```diff
@@ -120,38 +169,212 @@ def pretransform_single(self, value: RDKitMol) -> OptionalMol:
+        if any(
+            element not in self.allowed_element_numbers for element in elements_counter
+        ):
             return InvalidInstance(
-                self.uuid,
-                f"Molecule contains following forbidden elements: {forbidden_elements}",
-                self.name,
+                self.uuid, "Molecule contains forbidden chemical element.", self.name
             )
+        for element, (min_count, max_count) in self.allowed_element_numbers.items():
+            count = elements_counter[element]
+            if (min_count is not None and count < min_count) or (
+                max_count is not None and count > max_count
+            ):
+                return InvalidInstance(
+                    self.uuid,
+                    f"Molecule contains forbidden number of element {element}.",
+                    self.name,
+                )
         return value
 
 
-class MixtureFilter(_MolToMolPipelineElement):
-    """MolToMol which removes molecules composed of multiple fragments."""
+class SmartsFilter(_BasePatternsFilter):
+    """Filter to keep or remove molecules based on SMARTS patterns.
+
+    Notes
+    -----
+    There are four possible scenarios:
+        - mode = "any" & keep_matches = True: Needs to match at least one filter element.
+        - mode = "any" & keep_matches = False: Must not match any filter element.
+        - mode = "all" & keep_matches = True: Needs to match all filter elements.
+        - mode = "all" & keep_matches = False: Must not match all filter elements.
+    """
+
+    def _pattern_to_mol(self, pattern: str) -> RDKitMol:
+        """Convert SMARTS pattern to RDKit molecule.
+
+        Parameters
+        ----------
+        pattern: str
+            SMARTS pattern to convert.
+
+        Returns
+        -------
+        RDKitMol
+            RDKit molecule.
+        """
+        return Chem.MolFromSmarts(pattern)
+
 
-    def __int__(
+class SmilesFilter(_BasePatternsFilter):
+    """Filter to keep or remove molecules based on SMILES patterns.
+
+    Notes
+    -----
+    There are four possible scenarios:
+        - mode = "any" & keep_matches = True: Needs to match at least one filter element.
+        - mode = "any" & keep_matches = False: Must not match any filter element.
+        - mode = "all" & keep_matches = True: Needs to match all filter elements.
+        - mode = "all" & keep_matches = False: Must not match all filter elements.
+    """
+
+    def _pattern_to_mol(self, pattern: str) -> RDKitMol:
+        """Convert SMILES pattern to RDKit molecule.
+
+        Parameters
+        ----------
+        pattern: str
+            SMILES pattern to convert.
+
+        Returns
+        -------
+        RDKitMol
+            RDKit molecule.
+        """
+        return Chem.MolFromSmiles(pattern)
+
+
+class ComplexFilter(_BaseKeepMatchesFilter):
+    """Filter to keep or remove molecules based on multiple filter elements.
```

Comment:
I think we need to check the get and set params of the subfilters.  
To ensure the correct passing of sub-parameters I am afraid that we would need to provide them as a tuple in the init:
(name, object): (lower, upper)
name: str
object: filter
lower: float| int
upper: float| int

Reply:
I think we should quickly discuss on monday - I am not onehundred percent sure, how this actually helps

Reply:
This would help during the get/set_params section, as the sklearn schema does not support giving parameters to objects based on their position in a list. I feel this will be needed to save/load the object in a json file. See comment to the test below.

Reply:
I believe I implemented this - I needed to include this in the init and store an additional attribute (also poping attributes of the base class when calling get_params). Please check this one time @c-w-feldmann , but at least test looks good. Also implemented the deep stuff. Want to point out, that the __ settings get quite long (in a pipeline e.g.: MultiElementFilter__pipeline_filter_elements__element_filter_1__add_hydrogens, sets the ComplexFilter (name MultiElementFilter, attribute pipeline_filter_elements (which is a list of tuple (name, obje)), object by name (element_filter_1), attribute add_hydrogens  to the defined value